### ISSUE TITLE
leverage ssh config (was : aliases for groups of hosts)

### ISSUE THREAD

**Issue by parmentelat** [parmentelat/apssh#3]
Following up on a side-track of issue #2.

@asher256

At this point I have no plan in parsing ssh’s config file because it would
- be an awful lot of work, and
- there is **a lot** of ssh config details that just don’t make sense to us

What I had in mind though, was the ability for users to define their own target files in, say, ~/.apssh

So you would for example create a file

cat ~/.apssh/desktops
root@desktop.testserver.org
root@desktop2.testserver.org

And from then on you could just run
apssh -t desktops blabla

As far as I am concerned, I really need a way to create aliases for **groups** of nodes, after all that is what `apssh` is all about in the first place

Implementation would be straightforward, I would primarily just need to add this search location when resolving targets. 

It might also make sense to allow a target file to contain target names that are not hostnames, but in turn target files

Would this kind of things be helpful in your case ?

**Comment by ghost** [comment]
I agree. Rewriting an SSH config parser doesn't make sense ;) . It would be a waste of time IMHO. If you really need to parse it you can use some Python "paramiko" functions.

The solution you provided is interesting (root@desktop2.testserver.org in ~/.apssh/desktops), however it is not enough to replace the ~/.ssh/config, because in ~/.ssh/config you can define different private keys for a group of hosts, ProxyCommand (useful for trusted hosts) and many other useful SSH options. 

What kind of library / backend do you use for apssh to run SSH commands? Do you use the SSH command or a library like libssh/paramiko? Can we add a custom options to ssh like -F for example with apssh?

**Comment by parmentelat** [comment]
`apssh` is entirely based on asyncssh, and on `asyncio` more generally
For this reason I didn't know much about `paramiko`, but at first - and quick - sight [it looks like this library can be easily leveraged](http://docs.paramiko.org/en/2.0/api/config.html) for parsing a config file in an ssh format

this indeed opens up a lot of possibilities - too many maybe :-) so I need to prioritize here.

In a first pass I suggest to add the following minimal features:
. look into `~/.apssh` (or something else if specified with `-D`) to locate target files, as I mentioned above
. opening the ssh config when found (or something else specified with `-F`), and using this for a handful of supported configuration options. This of course will necessarily be a small subset of what ssh supports. I understand you'd like to see at least the following tags to be used
- `Hostname`
- `User`

You seem to suggest there is an ssh config option to specify private keys on a host-per-host basis, but I could not spot this one; in any case please suggest other ssh config options that could be added to this batch.

If we go down this road, there is the question of the interdependance of these two config schemes.
In the case where a name is found both in `~/.apssh` and in the ssh config file, my proposal is for the `.apssh` source to take precedence.

Also both schemes could be turned off by specifying either `-F /dev/null` or `-D /dev/null`

How would that sound ?

**Comment by parmentelat** [comment]
For now in version 0.2.8 I have added the search in ~/.apssh (not overridable with -D btw at this point)
As soon as you send your feedback, and esp. about locating keys on a node-per-node basis, I can go on with this feature

**Comment by parmentelat** [comment]
Let me add that I'd really like to try and leverage the ProxyCommand defined in ssh's config as well

apssh supports 2-hops connections through a gateway, and clearly most of the time this data is captured in such ProxyCommand's 
Of course I far as I can tell this would be heuristics of some kind, as the ProxyCommand itself can be anything.. 

Comments/thoughts/hints are more than welcome

**Comment by parmentelat** [comment]
The ability to have groups has been around for a while already

I am keeping this open as a hint to use another library - paramiko ? - to read the contents of the ssh config and try to leverage that as much as possible
